### ISSUE TITLE
Testcafe timeouts delay test execution unnecessarily.

### ISSUE THREAD

**Issue by Raynos** [DevExpress/testcafe#4805]
### What is your Test Scenario?

I am using `testcafe` with electron to test my app.
 
### What is the Current behavior?

The tests take quite a long time.

```
~/optoolco/logs on jakev.dev*
$ time npr testcafe -q test/testcafe/index.ts 
 Running tests in:
 - Electron 7.1.5 / Linux 0.0

 Logs
 ✓ main UI components exist
 ✓ groups modal renders groups
 ✓ groups modal selection behavior
 ✓ groups modal caching behavior
 ✓ streams modal
 ✓ log events
 ✓ scrolling log events
 ✓ filtering log events
 ✓ sorting the event logs table


 9 passed (1m 57s)

real	2m2.688s
user	0m16.890s
sys	0m2.462s
```
 
### What is the Expected behavior?

The tests can finish a lot more quickly.

```
~/optoolco/logs on jakev.dev*
$ time npr testcafe -q test/testcafe/index.ts --assertion-timeout 100 --selector-timeout 100 --app-init-delay 0
 Running tests in:
 - Electron 7.1.5 / Linux 0.0

 Logs
 ✓ main UI components exist
 ✓ groups modal renders groups
 ✓ groups modal selection behavior
 ✓ groups modal caching behavior
 ✓ streams modal
 ✓ log events
 ✓ scrolling log events
 ✓ filtering log events
 ✓ sorting the event logs table


 9 passed (58s)

real	1m3.916s
user	0m15.729s
sys	0m2.229s
```
 
To me it looks like the timeouts are too aggressive by default or there's something weird in my tests or apps that cause the timeout to always trigger.

**Comment by AlexSkorkin** [comment]
TestCafe uses the [Smart Assertion Query Mechanism](https://devexpress.github.io/testcafe/documentation/test-api/assertions/#smart-assertion-query-mechanism), which does not increase the test duration. TestCafe does not wait until the entire timeout exceeds. Instead, it checks for the assertion state within the timeout period multiple times until the actual property value in assertion is retrieved or the timeout is exceeded.

The same occurs with the [selector timeouts](https://devexpress.github.io/testcafe/documentation/test-api/selecting-page-elements/selectors/using-selectors.html#selector-timeout) - the selector is executed over and over again until it returns a DOM node or the timeout exceeds.

I recommend you profile your tests execution to check why it takes time ([What is the best way to profile javascript execution?](https://stackoverflow.com/questions/855126/what-is-the-best-way-to-profile-javascript-execution)).

**Comment by Raynos** [comment]
@AlexSkorkin 

My test execution time get's halfed from 2 minutes to 1 minute. This is not a performance issue in my app.

This is an issue with the selector timeout being. It's waiting the full 10s to execute a click (probably 6 times) which adds a while minute of latency.

Having a warning printed when the selector timeout triggers and the click succeeds anyway would greatly help with debugging.

**Comment by AlexSkorkin** [comment]
TestCafe does not wait until the entire timeout value is passed as it is checked periodically within the timeout value. If you have a scenario where we can see this behavior, please provide your sample, so that we can run it, and your complete test case that we can execute to examine the issue.

**Comment by Raynos** [comment]
Making a reproduction case is not easy. We are using testcafe + Electron.

**Comment by AlexSkorkin** [comment]
Thank you for your clarification.
Try the following workaround for your selector(s) and see if it makes any difference:
`Selector('smth').with({timeout: 0})`

**Comment by sarahncrowe** [comment]
> This is an issue with the selector timeout being. It's waiting the full 10s to execute a click (probably 6 times) which adds a while minute of latency.

TestCafe will wait for the selector to become visible and then if it times out, still clicks on it as like a "Hail Mary" (at least to my understanding). I'd suggest double checking that the selectors you're clicking on are visible.

Sorry if that's not helpful. We're not using electron but I've run into the same sort of issue, and the visibility was the cause on our end.

**Comment by Asuza** [comment]
I'm experiencing what I believe might be the same issue. Using `Selector` to target an element with an ID might cause TestCafe to not "find" the element until the timeout has finished, despite it being there on the page.

Using `Selector('smth').with({timeout: 0})` as suggested did help with the speed, but also made parts of the test flaky since it (presumably) needed more time than 0 to find the element. I wish I had hard evidence to give, but I've seen this occur since before the release of version 1.

If someone could point me in the right direction of where in the code TestCafe tries to find an element, I would be more than happy to try to debug the issue.

**Comment by Raynos** [comment]
I believe I might have found the cause of one of my issues.

I am using web components and custom elements a lot in my application. These elements have `display: inline` as that's the default CSS and then have a width & height of `0` ; This can cause testcafe to think that they are not visible when their children are clearly visible and have non-zero width & height.

**Comment by Asuza** [comment]
@Raynos your element does get found at the end of the timeout though, right? I would think your element wouldn't be found at all if it's related to the inline display and height/width you mentioned.

**Comment by soletan** [comment]
I'm currently investigating this (old) issue as it is happening in our tests very often.

The test

```
await t.click(buttons.filterVisible().withText(label));
```

is pending for 10s unnecessarily as the button is clearly visible and clickable.

I tried using a custom filter function instead of withtext() that's logging test results and it instantly reports true on the element in question. So I tested same case in a different way using TestCafe`s step debugger

```
const selector = buttons.filterVisible().withText(label);
await t.expect(selector.exists).ok();
await t.click(selector);
```

Executing this step by step, the `t.expect()` succeeds instantly, but the `t.click()` still takes 10s to click the visible and available button. IMHO this prooves that there is something odd inside the click() action. As a workaround, this is working for us for now:

```
const selector = buttons.filterVisible().withText(label);
await t.expect(selector.exists).ok();
await t.click(selector.with({timeout: 100}));
```
